src/imputation/tmi_imputation.py

```diff
@@ -0,0 +1,337 @@
+        trimmed_df[f"{target_variable}"].mean(),
+        len(trimmed_df),
+    ]
+
+    return dict_mean_growth_ratio
+
+
+def calculate_means(df, target_variable_list):
+
+    dfs_list = []
+
+    # Create an empty dict to store means
+    mean_dict = dict.fromkeys(target_variable_list)
+
+    copy_df = df.copy()
+
+    filtered_df = apply_filters(copy_df)
+
+    filtered_df = filtered_df[~(filtered_df["imp_class"].str.contains("nan"))]
+    filtered_df = filtered_df[filtered_df["formtype"] == "0001"]
+
+    # Group by imp_class
+    grp = filtered_df.groupby("imp_class")
+    class_keys = list(grp.groups.keys())
+
+    for var in target_variable_list:
+        for k in class_keys:
+            # Get subgroup dataframe
+            subgrp = grp.get_group(k)
+            # Sort by target_variable, df['employees'], reference
+            sorted_df = sort(var, subgrp)
+
+            # Add trimming threshold marker
+            trimcheck_df = trim_check(sorted_df)
+
+            # Apply trimming marker
+            trimmed_df = trim_bounds(trimcheck_df, var)
+
+            tr_df = trimmed_df.set_index("pre_index")
+
+            dfs_list.append(tr_df)
+            # Calculate mean and count # TODO: Rename this
+            means = get_mean_growth_ratio(trimmed_df, k, var)
+
+            # Update full dict with values
+            if mean_dict[var] is None:
+                mean_dict[var] = means
+            else:
+                mean_dict[var].update(means)
+
+    df = pd.concat(dfs_list)
+    df["qa_index"] = df.index
+    df = df.groupby(["pre_index"], as_index=False).first()
+
+    return mean_dict, df
+
+
+def tmi_imputation(df, target_variables, mean_dict):
+    
+    for var in target_variables:
+        df[f"{var}_imp_marker"] = "N/A"
+        df[f"{var}_imputed"] = df[var]
+
+    copy_df = df.copy()
+
+    filtered_df = filter_by_column_content(
+        copy_df, "status", ["Form sent out", "Check needed"]
+    )
+
+    filtered_df = filtered_df[~(filtered_df["imp_class"].str.contains("nan"))]
+
+    grp = filtered_df.groupby("imp_class")
+    class_keys = list(grp.groups.keys())
+
+    for var in target_variables:
+        for item in class_keys:
+
+            # Get grouped dataframe
+            subgrp = grp.get_group(item)
+
+            if f"{var}_{item}_mean and count" in mean_dict[var].keys():
+                # Replace nulls with means
+                subgrp[f"{var}_imputed"] = float(mean_dict[var][f"{var}_{item}_mean and count"][0])
+                subgrp[f"{var}_imp_marker"] = "TMI"
+
+            else:
+                subgrp[f"{var}_imputed"] = subgrp[var]
+                subgrp[f"{var}_imp_marker"] = "No mean found"
+
+            # Apply changes to copy_df
+            final_df = apply_to_original(subgrp, filtered_df)
+
+    final_df = apply_to_original(final_df, df)
+
+    return final_df
+
+
+def run_tmi(full_df, target_variables, sic_mapper):
```

Comment:
Could each step be commented and maybe link to the step in the methods for future ease of understanding